src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -793,40 +795,39 @@ JvmtiEnvBase::get_live_threads(JavaThread* current_thread, Handle group_hdl, jin
 
 jvmtiError
 JvmtiEnvBase::get_subgroups(JavaThread* current_thread, Handle group_hdl, jint *count_ptr, Handle **group_objs_p) {
-  ObjectLocker ol(group_hdl, current_thread);
 
-  int ngroups = java_lang_ThreadGroup::ngroups(group_hdl());
-  int nweaks = java_lang_ThreadGroup::nweaks(group_hdl());
+  // This call collects the strong and weak groups
+  JavaThread* THREAD = current_thread;
+  JavaValue result(T_OBJECT);
+  JavaCalls::call_virtual(&result,
+                          group_hdl,
+                          vmClasses::ThreadGroup_klass(),
+                          SymbolTable::new_permanent_symbol("subgroupsArray"),
+                          vmSymbols::void_threadgroup_array_signature(),
+                          THREAD);
+  if (HAS_PENDING_EXCEPTION) {
+    CLEAR_PENDING_EXCEPTION;
+    return JVMTI_ERROR_OUT_OF_MEMORY;
```

Comment:
Do we need to handle unexpected exceptions better, rather than just claiming they are OOME?

Reply:
I added a case for returning JVMTI_ERROR_INTERNAL also.  In the spec
https://docs.oracle.com/en/java/javase/11/docs/specs/jvmti.html#universal-error
The other universal errors are inapplicable (and already checked in the case of JVMTI_ERROR_INVALID_THREAD_GROUP and the wrong phase one (by the jvmti code wrapper).